### ISSUE TITLE
Moving a dynamic body with velocities?

### ISSUE THREAD

**Issue by marcell-baranyai** [NVIDIAGameWorks/PhysX#486]
Hello, this is not really a PhysX issue, I'd rather need technical advice that I couldn't find elsewhere so far.
I'm trying to integrate PhysX in my "high frequency" (1000Hz) vehicle simulation.

The basic idea was that I simulate my own vehicle's rigid body as usual, and when the time comes I send my simulated position into PhysX to handle collisions with the scene. I set the linear and angular velocities so my body would move to the desired location in one PhysX frame. And in case of collision I simply fetch back the transformation and new velocities and continue like nothing has happened.

This was the plan, and it works perfectly as long as there is no collision. Hitting static bodies are somewhat good, when the car rolls over... there are issues sometimes. When hitting or "stacking on" dynamic bodies, that's just catastrophic, not even close to my expectations.

I also tried accumulating forces/torques I calculated and added to the PhysX counterpart, but due to the low frequency it resulted in very bad jittering (the sum forces were scaled down to match the update rate differences). I tried with both forces and velocity changes, and in the later I dealt with mass and inertia tensor myself, but the results were the same.

So my questions are:

- Is this the preferred (and proven) way doing this sort of job, and the problem is with me?

- The ideal solution would be if I could send my forces in PhysX, but that would need smaller timesteps, at least for that specific body. Is there any "public" way of doing it?

Thanks in advance for any idea, suggestion or hint.

**Comment by kstorey-nvidia** [comment]
Applying forces to the bodies should be the right thing to do. This is how the PhysX vehicle component operates - it applies forces on the vehicle rigid body based on the tire, engine and suspension models. It does not directly set velocities. After forces are applied, the rigid bodies (including the vehicle body) are simulated and integrated by PhysX.

**Comment by marcell-baranyai** [comment]
Yes, I agree, adding forces would be the ideal way, but exactly this is my problem. I'd need to integrate the body at higher frequency, but running PhysX at 1000 fps would be an overkill :(

**Comment by gyeomannvidia** [comment]
The PhysX Vehicle SDK does something very close to what you want to achieve.  It's worth taking a look at its structure to get a flavour of the techniques involved.

The update of a physx vehicle has two distinct phases.  The first computes a force or velocity to apply to the physx actor, while the second forward integrates the physx  actor in the physx scene and accounts for rigid body collisions and for any constraints applied to the body.

The 1sts phase begins by caching the global pose of the physx actor.  It then substeps the computation of suspension and tire forces and forward integrates the global pose.  A really important pont is that this is done independently of the physx actor.  After N substeps, we apply the velocity to the physx actor. 

Pseudo-code as follows:

vehiclePosition = physxActor->getGlobalPose();
oldVehiclePosition  = vehiclePosition;
vehicleVelocity = physxActor->getLinearVelocity();
oldVehicleVelocity = vehicleVelocity;
for(i = 0; i < N; i++){
        suspensionCompression = computeSuspensionCompression(vehiclePosition);
        suspensionForce =  computeSuspensionForce(suspensionCompression);
        tireForce = computeTireForce(suspensionForce);
        vehicleVelocity +=  (suspensionForce + tireForce) * dt/N;
        vehiclePosition += vehicleVelocity * dt/N;
}

if(applyAsVelocity)
    physxActor->setLinearVelocity(vehicleVelocity);
else if(applyAsAveragedVelocity)
   physxActor->setLinearVelocity((vehiclePosition - oldVehiclePosition)/dt);
else if(applyAsAveragedAcceleration)
    physxActor->addForce((vehicleVelocity - oldVehicleVelocity)/dt, PxForceMode::eACCELERATION);

The second phase is a simple update of the physx scene that owns the physx actor.

physxScene->simulate(dt);

It is always best to avoid updating the global pose of the physx actor.  The key point here is that we don't know if the vehicle can actually get there or if it will hit something along the way.  A far better strategy is to apply either a velocity or a force to the physx actor and let physx work out what happens in the next dt.

In the example above, we set the linear velocity of the physx actor to  be either the linear velocity recorded after N substeps or an averaged velocity over the N substeps or an averaged acceleration of the N substeps.  Each of these has their own pros and cons.  As dt tends towards zero, however, all 3 choices converge.  Other alternatives, such as an instantaneous acceleration on the Nth substep, are also possible.   If the choices are wildly different, then the timestep is likely too large to offer fidelity or stability.   PhysX vehicles employ either velocity on Nth substep or averaged  acceleration over the N substeps and both seem to work quite well.  

One thing to bear in mind is whether you want the substepping phase to change the queriable state of the physx actor.  Applying an acceleration to the physx actor does not immediately change its queryable  state.  Applying a velocity, on the other hand, does.

One thing to note is that each phase must remain temporarally in sync.  If the substepping forward integrates P seconds, then the physx simulate step must also forward integrate P seconds.  It would be possible to break this rule by interpolating the output of the substepping and ensuring that substepping is always ahead of physx simulation time.  I've never tried this but it should be possible. 

We've used the substepping technique above for scenarios  that include Pacejka tires, which require 0.001 timesteps.  The physx scene was ticked at 0.016 and the vehicle model was ticked 16 times for each physx scene update. 

Hope this is useful,

Gordon

**Comment by marcell-baranyai** [comment]
Hi Gordon,
Thanks for the comprehensive information!
I write here a couple of consequences for future people with the same headache.

As I was browsing PhysX Vehicle, I realized that it's doing the same as I do, the only difference was that fixed sub step count, and due to my time accumulation, my vehicle was updated with different DTs. I don't know how much impact did it make, but in case of position based velocity change it was not consistent between frames.

I took 2 possible cases:
A. setVelocity(subStepSimVelocity);  // Setting the velocity I calculated during sub step integration
B. setVelocity((pos - oldPos) / dt);  // Setting the velocity based on body movement

And yes, both solution yields pretty much the same result, but as you mentioned both have pros/cons:
A: The calculated velocity fed in PhysX will not result in the same position that was integrated in sub-steps.
B: The position calculated by PhysX will match but the resulting velocities will be different.

In case B, fetching initial velocity from PhysX before update resulted in noticeable acceleration differences. A working solution was to only update from PhysX when the vehicle's body has collided in the last PhysX update. It has possible a drawback compared to case A, when a collision is persistent during driving.

In case A, I didn't notice any acceleration issues, and the small position mismatch was not disturbing during driving. But, due to lateral vibration of the tire model at low speed, some visible sliding artifact was introduced. Then again, it has an advantage if my vehicle is carrying other bodies.

I didn't try all possible combinations yet, but I feel that combining both method could lead to final victory.

Thanks again for the help!
